#! /bin/sh -e

# gdc-ice-lookup.dpatch by Iain Buclaw <ibuclaw@ubuntu.com>
# DP: Work around build on kfreebsd-amd64 until real problem is found.

dir=
if [ $# -eq 3 -a "$2" = '-d' ]; then
    pdir="-d $3"
    dir="$3/"
elif [ $# -ne 1 ]; then
    echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
    exit 1
fi
case "$1" in
    -patch)
        patch $pdir -f --no-backup-if-mismatch -p0 < $0
        #cd ${dir}gcc && autoconf
        ;;
    -unpatch)
        patch $pdir -f --no-backup-if-mismatch -R -p0 < $0
        #rm ${dir}gcc/configure
        ;;
    *)
        echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
        exit 1
esac
exit 0

--- gcc/ggc-page.c.orig	2007-11-01 22:20:29.000000000 +0000
+++ gcc/ggc-page.c	2010-03-14 21:24:16.000000000 +0000
@@ -1279,6 +1279,11 @@
   unsigned bit, word;
   unsigned long mask;
 
+  /* FIXME: p is not a valid page entry
+     (not allocated by the collector?), skip it for now. */
+  if (! ggc_allocated_p (p))
+    return 1;
+
   /* Look up the page on which the object is alloced.  If the object
      wasn't allocated by the collector, we'll probably die.  */
   entry = lookup_page_table_entry (p);
